Fix incorrect lxml usage

Installing lxml-stubs (`poetry add -G dev lxml-stubs`) finds some potentially incorrect usage of the lxml library within pyWeMo.

```
$ mypy pywemo
pywemo/exceptions.py:24: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/exceptions.py:24: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/exceptions.py:29: error: Statement is unreachable  [unreachable]
pywemo/ouimeaux_device/api/service.py:260: error: Value expression in dictionary comprehension has incompatible type "Optional[str]"; expected type "str"  [misc]
pywemo/ouimeaux_device/bridge.py:99: error: Unsupported operand types for in ("Union[str, None, Any]" and "Dict[str, Light]")  [operator]
pywemo/ouimeaux_device/bridge.py:99: error: Item "None" of "Optional[_Element]" has no attribute "text"  [union-attr]
pywemo/ouimeaux_device/bridge.py:102: error: Invalid index type "Union[str, None, Any]" for "Dict[str, Light]"; expected type "str"  [index]
pywemo/ouimeaux_device/bridge.py:105: error: Unsupported operand types for in ("Union[str, None, Any]" and "Dict[str, Group]")  [operator]
pywemo/ouimeaux_device/bridge.py:105: error: Item "None" of "Optional[_Element]" has no attribute "text"  [union-attr]
pywemo/ouimeaux_device/bridge.py:108: error: Invalid index type "Union[str, None, Any]" for "Dict[str, Group]"; expected type "str"  [index]
pywemo/ouimeaux_device/bridge.py:139: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:139: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:183: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:183: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:244: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:244: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:246: error: et.Element? has no attribute "find"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:248: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:249: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:320: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:320: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:324: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:325: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:326: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:327: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:328: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:329: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:330: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:345: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:345: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:347: error: et.Element? has no attribute "tag"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:348: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:350: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:353: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:457: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:457: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:460: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:462: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/ouimeaux_device/bridge.py:462: note: Perhaps you need "Callable[...]" or a callback protocol?
pywemo/ouimeaux_device/bridge.py:464: error: et.Element? has no attribute "tag"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:465: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:467: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:469: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:470: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/bridge.py:472: error: et.Element? has no attribute "findtext"  [attr-defined]
pywemo/ouimeaux_device/api/attributes.py:60: error: Key expression in dictionary comprehension has incompatible type "Optional[str]"; expected type "str"  [misc]
pywemo/ouimeaux_device/api/attributes.py:60: error: Value expression in dictionary comprehension has incompatible type "Optional[str]"; expected type "str"  [misc]
pywemo/ouimeaux_device/api/attributes.py:64: error: Argument 1 to "_is_int_or_float" has incompatible type "Optional[str]"; expected "str"  [arg-type]
pywemo/subscribe.py:344: error: et.Element? has no attribute "findall"  [attr-defined]
pywemo/subscribe.py:371: error: et.Element? has no attribute "find"  [attr-defined]
pywemo/subscribe.py:405: error: Function "lxml.etree.Element" is not valid as a type  [valid-type]
pywemo/subscribe.py:405: note: Perhaps you need "Callable[...]" or a callback protocol?
Found 43 errors in 5 files (checked 30 source files)
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,6 +37,7 @@ pylint = "^2.17.3"
 rstcheck = ">=3.3.1,<7.0.0"
 mypy = "^1.2.0"
 # testing and code coverage
+lxml-stubs = "^0.4.0"
 pytest = "^7.3.1"
 pytest-vcr = "1.0.2"
 types-requests = ">=2.0"

--- a/pywemo/exceptions.py
+++ b/pywemo/exceptions.py
@@ -21,7 +21,7 @@ class SOAPFault(ActionException):
     error_description: str = ""
 
     def __init__(
-        self, message: str = "", fault_element: et.Element | None = None
+        self, message: str = "", fault_element: et._Element | None = None
     ) -> None:
         """Initialize from a SOAP Fault lxml.etree Element."""
         details = ""
@@ -31,14 +31,13 @@ class SOAPFault(ActionException):
                 "/{urn:schemas-upnp-org:control-1-0}UPnPError/"
                 "{urn:schemas-upnp-org:control-1-0}"
             )
-            self.fault_code = fault_element.findtext("faultcode") or ""
-            self.fault_string = fault_element.findtext("faultstring") or ""
-            self.error_code = (
-                fault_element.findtext(f"{upnp_error_prefix}errorCode") or ""
+            self.fault_code = fault_element.findtext("faultcode", "")
+            self.fault_string = fault_element.findtext("faultstring", "")
+            self.error_code = fault_element.findtext(
+                f"{upnp_error_prefix}errorCode", ""
             )
-            self.error_description = (
-                fault_element.findtext(f"{upnp_error_prefix}errorDescription")
-                or ""
+            self.error_description = fault_element.findtext(
+                f"{upnp_error_prefix}errorDescription", ""
             )
             details = (
                 f" SOAP Fault {self.fault_code}:{self.fault_string}, "

--- a/pywemo/ouimeaux_device/api/attributes.py
+++ b/pywemo/ouimeaux_device/api/attributes.py
@@ -61,7 +61,10 @@ class AttributeDevice(Switch):
                 for attribute in et.fromstring(
                     xml_blob, parser=et.XMLParser(resolve_entities=False)
                 )
-                if len(attribute) >= 2 and _is_int_or_float(attribute[1].text)
+                if len(attribute) >= 2
+                and attribute[0].text is not None
+                and attribute[1].text is not None
+                and _is_int_or_float(attribute[1].text)
             }
         )
 

--- a/pywemo/ouimeaux_device/api/service.py
+++ b/pywemo/ouimeaux_device/api/service.py
@@ -257,7 +257,7 @@ class Action:
                         fault_element=response_element,
                     )
                 return {
-                    response_item.tag: response_item.text
+                    response_item.tag: response_item.text or ""
                     for response_item in response_element
                 }
 

--- a/pywemo/ouimeaux_device/bridge.py
+++ b/pywemo/ouimeaux_device/bridge.py
@@ -9,6 +9,7 @@ from typing import Any, TypedDict
 from lxml import etree as et
 
 from ..color import ColorXY, get_profiles, limit_to_gamut
+from ..exceptions import InvalidSchemaError
 from . import Device
 from .api.service import RequiredService
 
@@ -96,16 +97,24 @@ class Bridge(Device):
             )
 
             for light in end_device_list.iter("DeviceInfo"):
-                if (uniqueID := light.find("DeviceID").text) in self.lights:
-                    self.lights[uniqueID].update_state(light)
+                if not (device_id := light.findtext("DeviceID")):
+                    raise InvalidSchemaError(
+                        f"DeviceID missing: {et.tostring(light).decode()}"
+                    )
+                if device_id in self.lights:
+                    self.lights[device_id].update_state(light)
                 else:
-                    self.lights[uniqueID] = Light(self, light)
+                    self.lights[device_id] = Light(self, light)
 
             for group in end_device_list.iter("GroupInfo"):
-                if (uniqueID := group.find("GroupID").text) in self.groups:
-                    self.groups[uniqueID].update_state(group)
+                if not (group_id := group.findtext("GroupID")):
+                    raise InvalidSchemaError(
+                        f"GroupID missing: {et.tostring(group).decode()}"
+                    )
+                if group_id in self.groups:
+                    self.groups[group_id].update_state(group)
                 else:
-                    self.groups[uniqueID] = Group(self, group)
+                    self.groups[group_id] = Group(self, group)
 
         return self.lights, self.groups
 
@@ -136,7 +145,7 @@ class Bridge(Device):
             return False
         return super().subscription_update(_type, _param)
 
-    def bridge_getdevicestatus(self, deviceid: str) -> et.Element | None:
+    def bridge_getdevicestatus(self, deviceid: str) -> et._Element | None:
         """Return the list of device statuses for the bridge's lights."""
         status_list = self.bridge.GetDeviceStatus(DeviceIDs=deviceid)
         device_status_list_xml = status_list.get("DeviceStatusList")
@@ -180,20 +189,20 @@ class DeviceState(TypedDict, total=False):
 class LinkedDevice:
     """Representation of a device connected to the bridge."""
 
-    def __init__(self, bridge: Bridge, info: et.Element) -> None:
+    def __init__(self, bridge: Bridge, info: et._Element) -> None:
         """Create a Linked Device."""
-        self.bridge = bridge
-        self.host = self.bridge.host
-        self.port = self.bridge.port
-        self.name = ""
+        self.bridge: Bridge = bridge
+        self.host: str = self.bridge.host
+        self.port: int = self.bridge.port
+        self.name: str = ""
         self.state: DeviceState = {}
         self.capabilities: list[str] = []
         self._values: list[str] = []
         self.update_state(info)
         self._last_err: dict[str, str] = {}
-        self.mac = self.bridge.mac
-        self.serial_number = self.bridge.serial_number
-        self.uniqueID = ""
+        self.mac: str = self.bridge.mac
+        self.serial_number: str = self.bridge.serial_number
+        self.uniqueID: str = ""
 
     def get_state(self, force_update: bool = False) -> DeviceState:
         """Return the status of the device."""
@@ -241,10 +250,11 @@ class LinkedDevice:
             colorx, colory = colorx / 65535.0, colory / 65535.0
             self.state["color_xy"] = colorx, colory
 
-    def subscription_update(self, state_event: et.Element) -> bool:
+    def subscription_update(self, state_event: et._Element) -> bool:
         """Update the light values due to a subscription update event."""
-        device_id = state_event.find("DeviceID")
-        if device_id.get("available", "YES").upper() == "YES":
+        if (
+            device_id := state_event.find("DeviceID")
+        ) is None or device_id.get("available", "YES").upper() == "YES":
             capability = state_event.findtext("CapabilityId")
             value = state_event.findtext("Value")
         else:
@@ -317,17 +327,17 @@ class LinkedDevice:
 class Light(LinkedDevice):
     """Representation of a Light connected to the Bridge."""
 
-    def __init__(self, bridge: Bridge, info: et.Element) -> None:
+    def __init__(self, bridge: Bridge, info: et._Element) -> None:
         """Create a Light device."""
         super().__init__(bridge, info)
 
-        self.device_index = info.findtext("DeviceIndex")
-        self.uniqueID = info.findtext("DeviceID")
-        self.iconvalue = info.findtext("IconVersion")
-        self.firmware = info.findtext("FirmwareVersion")
-        self.manufacturer = info.findtext("Manufacturer")
-        self.model = info.findtext("ModelCode")
-        self.certified = info.findtext("WeMoCertified")
+        self.device_index: str = info.findtext("DeviceIndex", "")
+        self.uniqueID: str = info.findtext("DeviceID", "")
+        self.iconvalue: str = info.findtext("IconVersion", "")
+        self.firmware: str = info.findtext("FirmwareVersion", "")
+        self.manufacturer: str = info.findtext("Manufacturer", "")
+        self.model: str = info.findtext("ModelCode", "")
+        self.certified: str = info.findtext("WeMoCertified", "")
 
         self.temperature_range, self.gamut = get_profiles(self.model)
         self._pending: dict[str, Any] = {}
@@ -342,10 +352,10 @@ class Light(LinkedDevice):
 
         return self
 
-    def update_state(self, status: et.Element) -> None:
+    def update_state(self, status: et._Element) -> None:
         """Update the device state."""
         if status.tag == "DeviceInfo":
-            self.name = status.findtext("FriendlyName")
+            self.name = status.findtext("FriendlyName", "")
 
         capabilities = status.findtext("CapabilityIDs") or status.findtext(
             "CapabilityID"
@@ -454,15 +464,15 @@ class Light(LinkedDevice):
 class Group(LinkedDevice):
     """Representation of a Group of lights connected to the Bridge."""
 
-    def __init__(self, bridge: Bridge, info: et.Element) -> None:
+    def __init__(self, bridge: Bridge, info: et._Element) -> None:
         """Create a Group device."""
         super().__init__(bridge, info)
-        self.uniqueID = info.findtext("GroupID")
+        self.uniqueID: str = info.findtext("GroupID", "")
 
-    def update_state(self, status: et.Element) -> None:
+    def update_state(self, status: et._Element) -> None:
         """Update the device state."""
         if status.tag == "GroupInfo":
-            self.name = status.findtext("GroupName")
+            self.name = status.findtext("GroupName", "")
 
         capabilities = status.findtext(
             "GroupCapabilityIDs"

--- a/pywemo/subscribe.py
+++ b/pywemo/subscribe.py
@@ -343,8 +343,12 @@ class RequestHandler(BaseHTTPRequestHandler):
             doc = self._get_xml_from_http_body()
             for propnode in doc.findall(f"./{NS}property"):
                 for property_ in list(propnode):
-                    text = property_.text or ""
-                    outer.event(device, property_.tag, text, path=self.path)
+                    outer.event(
+                        device,
+                        property_.tag,
+                        property_.text or "",
+                        path=self.path,
+                    )
 
         self._send_response(200, RESPONSE_SUCCESS)
 
@@ -368,10 +372,8 @@ class RequestHandler(BaseHTTPRequestHandler):
                 )
             else:
                 doc = self._get_xml_from_http_body()
-                binary_state = doc.find(".//BinaryState")
-                if binary_state is not None:
-                    text = binary_state.text
-                    outer.event(device, EVENT_TYPE_LONG_PRESS, text)
+                if binary_state := doc.findtext(".//BinaryState"):
+                    outer.event(device, EVENT_TYPE_LONG_PRESS, binary_state)
             self._send_response(200, RESPONSE_SUCCESS)
         else:
             self._send_response(404, RESPONSE_NOT_FOUND)
@@ -402,7 +404,7 @@ class RequestHandler(BaseHTTPRequestHandler):
         if body:
             self.wfile.write(body.encode("UTF-8"))
 
-    def _get_xml_from_http_body(self) -> et.Element:
+    def _get_xml_from_http_body(self) -> et._Element:
         """Build the element tree root from the body of the http request."""
         content_len = int(self.headers.get("content-length", 0))
         data = self.rfile.read(content_len)

